### DISCUSSION TITLE
AS4 User Message Properties

### DISCUSSION THREAD

**Discussion by Nagendra-Naidu1629** [phax/phase4#78] [Q&A]
Dear, We are having an issue while parsing an incoming AS4 message when the "MessageProperties" is sent as a empty element.
We are using CEF profile and in the validator it is made as mandatory but as per the specification it is an optional element.
Below is an extract of the incoming message,

```xml
<eb:CollaborationInfo>
		<eb:Service type="test:service">testService</eb:Service>
		<eb:Action>testAction</eb:Action>
		<eb:ConversationId>test@domibus.eu
		</eb:ConversationId>
</eb:CollaborationInfo>
<eb:MessageProperties />
```

This message fails parsing and validation due to empty <eb:MessageProperties />

Please help.

**Comment by phax** [comment]
This is, because the `<MessageProperties>` element, as defined in the XML Schema, may not be empty. It MUST contain at least one `<Property>` element to be compliant.

If you want no Message properties, just avoid that empty element `<eb:MessageProperties />` completely

Edit: see https://github.com/phax/phase4/blob/master/phase4-lib/src/main/resources/schemas/ebms-header-3_0-200704.xsd for the full XML Schema

**Reply by Nagendra-Naidu1629** [reply]
Thank you for the answer. Profile CEF is having a validator which check for this as a mandatory element, please suggest.

CEFCompatibilityValidator class - extract of this below.
```java
// Check if originalSender and finalRecipient are present
        // Since these two properties are mandatory
        final Ebms3MessageProperties aMessageProperties = aUserMsg.getMessageProperties ();
        if (aMessageProperties == null)
          aErrorList.add (_createError ("MessageProperties is missing but 'originalSender' and 'finalRecipient' properties are required"));
        else
        {
          final List <Ebms3Property> aProps = aMessageProperties.getProperty ();
          if (aProps.isEmpty ())
            aErrorList.add (_createError ("MessageProperties/Property must not be empty"));
          else
          {
            String sOriginalSenderC1 = null;
            String sFinalRecipientC4 = null;

            for (final Ebms3Property sProperty : aProps)
            {
              if (sProperty.getName ().equals (CAS4.ORIGINAL_SENDER))
                sOriginalSenderC1 = sProperty.getValue ();
              else
                if (sProperty.getName ().equals (CAS4.FINAL_RECIPIENT))
                  sFinalRecipientC4 = sProperty.getValue ();
            }

            if (StringHelper.hasNoText (sOriginalSenderC1))
              aErrorList.add (_createError ("MessageProperties/Property '" +
                                            CAS4.ORIGINAL_SENDER +
                                            "' property is empty or not existant but mandatory"));
            if (StringHelper.hasNoText (sFinalRecipientC4))
              aErrorList.add (_createError ("MessageProperties/Property '" +
                                            CAS4.FINAL_RECIPIENT +
                                            "' property is empty or not existant but mandatory"));
          }
        }
```

**Reply by phax** [reply]
Ah okay, now I got it.

According to https://ec.europa.eu/digital-building-blocks/wikis/display/CEFDIGITAL/eDelivery+AS4+-+1.15, when using the four-corner model, the `originalSender` and `finalRecipient` properties must be present (see chapter 4.1.2).

Are you using phase4 in a non-four-corner model? Than I would need to ammend the Validator code

**Reply by Nagendra-Naidu1629** [reply]
Yes we are not using four corner model.

**Reply by phax** [reply]
Okay thanks - I will ammend the code

**Reply by Nagendra-Naidu1629** [reply]
Thank you

**Reply by phax** [reply]
Issue #79 it is

**Reply by phax** [reply]
This will be fixed in version 1.3.6

**Reply by Nagendra-Naidu1629** [reply]
Thanks a lot for the quick response and update.

**Reply by phax** [reply]
The new versin will be released this week
